Support set schema inference function in python

OYCN opened the pull request:
### Description

Support set schema inference callback function for custion OpSchema before register in python.

example:

```
def func(ctx: onnx.shape_inference.InferenceContext):
    # get some info
    assert ctx.get_num_inputs() == 2
    value = ctx.get_input_type(0)
    ...
    # get or create output proto object
    output = ctx.get_output_type(0)
    # set type or shape
    ...
    # set the result proto
    ctx.set_output_type(0, output)

schema.set_type_and_shape_inference_function(func)
```

### Note

Depends on https://github.com/onnx/onnx/pull/5906

### Motivation and Context

Follow up of https://github.com/onnx/onnx/issues/5019

OYCN commented:
We're spending a considerable amount of code on passing proto objects between C++ and Python. However, the [repository](https://github.com/pybind/pybind11_protobuf) seems well-prepared for this feature.

OYCN commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +632,85 @@
+  inference_ctx.def("get_num_inputs", &InferenceContext::getNumInputs);
+  inference_ctx.def("has_input", &InferenceContext::hasInput);
+  inference_ctx.def("__get_input_type", [](InferenceContext* ctx, size_t index) {
+    auto type = ctx->getInputType(index);
```
done

OYCN commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +632,85 @@
+    return py::bytes(data);
+  });
+  inference_ctx.def("__get_input_data", [](InferenceContext* ctx, size_t index) {
+    auto tensor = ctx->getInputData(index);
```
done

OYCN commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +632,85 @@
+  py::class_<InferenceContext> inference_ctx(shape_inference, "InferenceContext", "Inference context");
+
+  inference_ctx.def("__get_attribute", [](InferenceContext* ctx, std::string name) {
+    auto attr = ctx->getAttribute(name);
```
done

OYCN commented:
Hi @justinchuby ,

In this PR, we can implement shape inference on the Python side, similar to how it's done on the C++ side. If you have any suggestions for this implementation, I'm open to making adjustments accordingly.

justinchuby commented:
Thank you! Is it ready to be reviewed?

OYCN commented:
> Thank you! Is it ready to be reviewed?

Yes, I have removed the 'WIP' prefix from the title. Please feel free to leave any comments.  :D

xadupre commented on onnx/test/shape_inference_test.py:
```diff
@@ -10172,6 +10172,103 @@ def test_check_type_when_schema_has_empty_io(self):
+                )
+
+            assert ctx.get_num_inputs() == 2
+            in0 = ctx.get_input_type(0)
```
My concern with this is it goes through serialization to access the type information. It is not really efficient. I would change the API so that it does not return a TypeProto but the type and the shape as regular python objects.

gramalingam replied:
I think I agree with Xavier, but I am a bit confused also. I see the method implementation serializes proto values to string and returns them. We could just return a pointer to the C++ Proto object (wrapped as a Python object). Is that your suggestion Xavier?

gramalingam replied:
It looks like you are using pybind11's protobuf conversion (between C++ and python) above? Should this example be updated to match that? I guess it is still doing the same thing, but at least we don't need extra-code to manage the serialization/deserialization?

It seems ok to me, in the sense this is what we currently have for Python - C++ inter-op anyway (that is, serialize/deserialize). Until we come up with a better solution, this seems fine.

xadupre replied:
Let's keep it as it is. We can improve later.

xadupre commented on onnx/test/shape_inference_test.py:
```diff
@@ -10172,6 +10172,103 @@ def test_check_type_when_schema_has_empty_io(self):
+                out.tensor_type.shape.dim.add().dim_value = N
+                out.tensor_type.shape.dim.add().dim_value = La * Lb
+                out.tensor_type.shape.dim.add().dim_value = out_len[i]
+                ctx.set_output_type(i, out)
```
Same comment here, we should avoid serialization with something like ``set_output_type_and_shape(in0_type, (N, La*Lb, out_lin[i])``. The type is created on C++ side, there is no serialization and it would be more efficient.

OYCN replied:
Thanks, done.

xadupre commented:
I would be a nice feature to have.

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -114,6 +116,9 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
   );
 
+  // Avoid Segmentation fault if we not free the python function in Custom Schema
+  onnx_cpp2py_export.add_object("_cleanup", py::capsule([] { OpSchemaRegistry::OpSchemaDeregisterAll(); }));
```
Can you clarify when this gets invoked?

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +636,127 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+    }
+    auto attr = ctx->getAttribute(name);
+    if (attr == nullptr) {
+      fail_shape_inference("Internal error: `attr` is nullptr in `__get_attribute`");
```
This is not an error situation. In C++, this single method plays a dual role, both to check if an attribute is present, and to get its value. Is it possible to return `None` in this situation?

gramalingam replied:
Or, just a return a C++ pointer (using pybind's conventions ... it must have some standard approach).

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +636,127 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+    }
+    auto tensor = ctx->getInputData(index);
+    if (tensor == nullptr) {
+      fail_shape_inference("Internal error: `tensor` is nullptr in `__get_input_data`");
```
Again, not an error. Signifies no static-constant-value is known for this input. Best to map this to `None` or some similar thing

gramalingam commented on onnx/cpp2py_export.cc:
```diff
@@ -625,6 +636,127 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+    }
+    auto stensor = ctx->getInputSparseData(index);
+    if (stensor == nullptr) {
+      fail_shape_inference("Internal error: `stensor` is nullptr in `__get_input_sparse_data`");
```
As above. Map to `None`?

gramalingam commented:
Thanks for creating the PR! It would be great to add this functionality. My comments above.

xadupre commented:
I would do this in a different way. I think we should export in python all functions doing shape inference for every operator. We should have a python version of the algorithm doing shape inference. That way, adding new shape inference function would be easier and we could do local shape inference. This feature is needed when building a model. Shapes can be used to write efficient sequences of onnx operators.

OYCN commented:
Hi Xavier,

After thinking about it for a while, please correct me if my understanding is wrong. Implementing a Python version of shape inference may not eliminate the need to pass ONNX protos between C++ and Python. In the current architecture, if we want to obtain any proto in the shape inference function, we need to export `InferenceContext` from C++ to Python. Perhaps a more efficient implementation for interacting with information carried by proto is key to solving the problem. This could be done using native Python types (though it’s challenging to deliver graph protos), type pointers, or other methods.

Best Regards.

gramalingam commented:
The code-freeze for 1.17 is coming up soon. We may need to move this to a later release.

OYCN commented:
Hi @gramalingam

Sorry for the delay. Please let me briefly describe the changes I made since my last commit.

 - Because `pybind_protobuf` depends on `absl::strings`, I moved the `FetchContent_*` to the root scope in CMake and controlled them by the `Build_*` variable.
 - Added a check for the protobuf minor version, ensuring that it is at least x.22.y. This restriction comes from `pybind_protobuf`, so some CI will fatal. I think this is a bit radical, but I'm having some difficulty trying to compile protobuf as a fallback option.

It looks like there are plans to upgrade protobuf? (https://github.com/onnx/onnx/pull/6167)

andife commented:
@OYCN https://github.com/onnx/onnx/pull/6725 Minimum protobuf version was upgraded to "protobuf version of 25.1" for upcoming 1.18 release

justinchuby commented:
Sorry for the delayed response. We are interested in getting this PR in if you are still interested

OYCN commented:
Hi @justinchuby , 

For 'Ubuntu External' CI, the version of protobuf is relatively low (3.21.12). I can't determine what to do (upgrade or force build locally).
Because the pr depends on 'pybind11_protobuf', 'protobuf-lite' may no longer be used. At the same time, I found that only Windows ci used protobuf-lite, is there any reason for this.

justinchuby commented:
I think it’s just using different combinations of configs in CI. We build the python wheel using protobuf lite iirc

justinchuby commented:
I think updating protobuf version and moving to using protobuf for the python build is reasonable.

andife commented:
> I think updating protobuf version and moving to using protobuf for the python build is reasonable, if needed

Which protobuf version would be needed?

OYCN commented:
Consider possible compatibility issues with zero copy (In fact, we have not enable this feature at present based on pybind11_protobuf), I removed pybind11_protobuf in the latest commit and add the object transport based on serialization and deserialization. Further optimization may be carried out in the future. If anyone has any suggestions, please feel free to comment.

For new types, it is expected to use [these](https://github.com/onnx/onnx/blob/f9e5b0cf0cfd288624e308263045a47dea4fd7b2/onnx/cpp2py_export.cc#L76-L80) to add support. And it will be check at build-time (Only for [Message/Lite Type](https://github.com/onnx/onnx/blob/f9e5b0cf0cfd288624e308263045a47dea4fd7b2/onnx/cpp2py_export.cc#L42-L49))

OYCN commented:
Maybe I should add more unit tests.

OYCN commented:
Hi @justinchuby , the PR has been ready for review. If you have any suggestion, please feel free to comment.

justinchuby commented on onnx/cpp2py_export.cc:
```diff
@@ -634,6 +715,48 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+        auto inputTypesTuple = ConvertPyObjToPtr<ONNX_NAMESPACE::TypeProto>(inputTypesObj);
+        auto inputDataTuple = ConvertPyObjToPtr<ONNX_NAMESPACE::TensorProto>(inputDataObj);
+        auto ret = self.doInferencing(std::get<1>(inputTypesTuple), std::get<1>(inputDataTuple));
+        std::vector<py::object> retObj(ret.size());
```
```suggestion
        std::vector<py::object> ret_obj(ret.size());
```

nit: Use snake_case for variable names

justinchuby commented:
I think the main changes look good! Sorry for the delay. Just some minor nits

justinchuby commented on onnx/cpp2py_export.cc:
```diff
@@ -26,6 +28,57 @@
+template <typename T>
+struct PythonProtoTypeMap {};
+
+#define DEFINE_PROTO_TYPE_MAP(_ProtoType, PY_MODULE_NAME, PY_TYPE_NAME)                             \
```
For this part, is it possible to do serialization on the python side? Similar to https://github.com/onnx/onnx/blob/da8d7c1b404ce673c21a41a39ec17df91c80ef17/onnx/shape_inference.py#L48

Would that be cleaner?

justinchuby commented:
cc @gramalingam @xadupre

xadupre commented:
It would be great to add an example in the documentation. Let's merge this one but a standalone example would be great.
